runtime/src/main/kotlin/kotlin/native/concurrent/MutableData.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+import kotlinx.cinterop.*
+
+@SymbolName("Kotlin_Any_share")
+external private fun Any.share()
+
+@SymbolName("Kotlin_CPointer_CopyMemory")
+external private fun CopyMemory(to: COpaquePointer?, from: COpaquePointer?, count: Int)
+
+/**
+ * Mutable concurrently accessible data buffer. Could be accessed from several workers simulteniously.
+ */
+@Frozen
+public class MutableData {
+    init {
+        // Instance of MutableData is shared.
+        share()
+    }
+
+    private var buffer = ByteArray(16).apply { share() }
+    private var size_ = 0
+    private val lock = Lock()
+
+    private fun resizeDataLocked(newSize: Int): Int {
+        if (newSize > buffer.size) {
+            val actualSize = maxOf(buffer.size * 3 / 2 + 1, newSize)
+            val newBuffer = ByteArray(actualSize)
+            buffer.copyRangeTo(newBuffer, 0, size, 0)
+            newBuffer.share()
+            buffer = newBuffer
+        }
+        val position = size
+        size_ = newSize
+        return position
+    }
+
+    /**
+     * Current data size, may concurrently change later on.
+     */
+    public val size: Int
+        get() = size_
+
+    /**
+     * Reset the data buffer, makings its size 0.
+     */
+    public fun reset() = locked(lock) {
+        size_ = 0
+    }
+
+    /**
+     * Appends data to the buffer.
+     */
+    public fun append(data: MutableData) = locked(lock) {
+        val toCopy = data.size
+        val where = resizeDataLocked(size + toCopy)
+        data.copyRangeTo(buffer, 0, size, where)
+    }
+
+    /**
+     * Appends byte array to the buffer.
+     */
+    public fun append(data: ByteArray, start: Int = 0, count: Int = data.size - start): Unit = locked(lock) {
+        val where = resizeDataLocked(this.size + count)
+        data.copyRangeTo(buffer, start, count, where)
+    }
+
+    /**
+     * Appends C data to the buffer.
+     */
+    public fun append(data: COpaquePointer?, count: Int): Unit = locked(lock) {
```

Comment:
why is it nullable if a call with null doesn't do anything?

Reply:
Just for the ease of use with C/Objective-C APIs.

Reply:
I bet suppressing errors is not the ease we are looking to.

Reply:
Hmm, it really depends on what kind of operations we expect here. IMO, null are legit common case here, which shall be tolerated.